src/thunderbots/software/network_input/filter/ball_filter.cpp

```diff
@@ -1,35 +1,214 @@
 
-    previous_ball_readings.push_back(ball_velocity);
+    if(relative_error_inv < relative_error) {
+        best_fit_line_vector = best_fit_line_vector_inv;
+        p1 = p1_inv;
+        p2 = p2_inv;
+    }
+
+
+    // Take the position of the most recent ball position and project it onto the line of best fit. We do this because
+    // we assume the ball must be travelling along its velocity vector, and this allows us to return more stable
+    // position values since the line of best fit is less likely to fluctuate compared to the raw position of a ball
+    // detection
+    SSLBallDetection latest_ball_detection = ball_detections.front();
+    Point filtered_ball_position = closestPointOnLine(latest_ball_detection.position, p1, p2);
 
-    auto average_ball_velocity = Vector(0, 0);
-    for (auto v : previous_ball_readings)
-    {
-        average_ball_velocity = average_ball_velocity + v;
+    // Calculate the velocity of the ball using the average of the velocities calculated from the values currently
+    // in the detection buffer
+    auto velocities = calculateBallVelocities(ball_detections);
+    Vector average_velocity = Vector(0, 0);
+    for(const auto& v : velocities) {
+        average_velocity = average_velocity + v;
     }
-    average_ball_velocity = average_ball_velocity.norm(average_ball_velocity.len() /
-                                                       previous_ball_readings.size());
+    average_velocity = average_velocity.norm(average_velocity.len() / velocities.size());
+    double average_velocity_magnitude = average_velocity.len();
 
-    return Ball{filtered_detection.position, average_ball_velocity,
-                filtered_detection.timestamp};
+    // Project the velocity so it also lies along the line of best fit calculated above. Again, this gives more stable
+    // values because the line of best fit is more robust to fluctiations in ball position, so will not vary as much
+    // as using the "raw" ball velocity
+    Vector vel_along_line = average_velocity.project(best_fit_line_vector);
+    Vector filtered_velocity = vel_along_line.norm(average_velocity_magnitude);
+
+    Ball ball(filtered_ball_position, filtered_velocity, latest_ball_detection.timestamp);
+    return ball;
+}
+
+std::optional<Ball> BallFilter::getFilteredData(const std::vector<SSLBallDetection>& new_ball_detections, const Field& field)
+{
+    addNewDetectionsToBuffer(new_ball_detections, field);
+
+    if(ball_detection_buffer.size() >= 2) {
+        return getLinearRegressionPositionAndVelocity(ball_detection_buffer);
+    }else if(ball_detection_buffer.size() == 1) {
+        // If there is only 1 entry in the buffer, we can't calculate a velocity so
+        // just set it to 0
+        Ball filtered_ball = Ball(ball_detection_buffer.front().position, Vector(0, 0), ball_detection_buffer.front().timestamp);
+        return filtered_ball;
+    }else {
+        return std::nullopt;
+    }
 }
```

Comment:
In general, excellent job commenting in this file! I found it pretty easy to follow.